uPortal-api/uPortal-api-rest/src/main/java/org/apereo/portal/rest/layout/MarketplaceEntryRating.java

```diff
@@ -14,15 +14,56 @@
  */
 package org.apereo.portal.rest.layout;
 
+import java.util.Date;
+
 public class MarketplaceEntryRating {
 
+    public MarketplaceEntryRating() {}
+
     public MarketplaceEntryRating(int rating, String review) {
         this.rating = rating;
         this.review = review;
     }
 
     private int rating;
     private String review;
+    private String user;
+    private String portletName;
+    private Date ratingDate;
```

Comment:
:question: Is the column automatically created?
:question: How are missing dates handled?

Reply:
I don't understand your questions - the values will be null if not provided.

Reply:
I'm working to understand how this should be versioned, if this will be [SemVer minor or SemVer major](https://semver.org/spec/v2.0.0.html).

The first question has to do with if this code is run on an existing database, from before this column is added, does HSQL automatically add the column or does a manual migration have to be run.

For the second, when legacy data is brought in with date being `null`, does the code have a graceful fllaback? or will it NPE?

Reply:
I'll have to test this update but I think it will require a manual migration.

I'll run the current branch locally and do some ratings of portlets.

How do I run my  branch locally with uPortal-Start?  

I don't think I introduced any code that depends on ratingDate having a value but I will double check.

Reply:
>How do I run my branch locally with uPortal-Start?

From the branch on uPortal `./gradlew install`, this will build uPortal and create a snapshot on maven local.
Then head over to uPortal-start update `gradle.properties` to have `uPortalVersion=5.0.3-SNAPSHOT`.
Then run uPortal-start as usual, it will pull from the mavel local version.

:notebook: further updates to uPortal branch will require running `./gradlew install` on the uPortal side again.

Reply:
Here is how I tested

Ran uPortal-start with 5.0.1 version of uPortal

Created a few ratings as student and staff user

Built and installed locally my forked version of uPortal that includes adding RatingDate and new REST end points for market place controller.

Changed uPortal-start to use 5.0.3-SNAPSHOT and redeployed to Tomcat

Viewing any previous ratings or trying to create new ratings fails with SQL Exception due to missing RATINGDATE column in up_portlet_rating table

Added RATINGDATE column to up_portlet_rating table in HSQL database running locally using Squirrel SQL Client

Now can view previous ratings and can create new ratings

Tested new end points (e.g. http://localhost:8080/uPortal/api/marketplace/calendar/30/getRatings and http://localhost:8080/uPortal/api/marketplace/30/getRatings) and verified they work correctly.

Tested previous end points (e.g. http://localhost:8080/uPortal/api/marketplace/snooper/getRating and http://localhost:8080/uPortal/api/v5-0/marketplace/weather/ratings) and verified they work as before.

No Null Pointer Exceptions were generated when accessing ratings that were done previously and do not have a RATINGDATE value.

What is uPortal's process for updating the database schema to support a new release?

Do you feel this must be a major semantic version update (e.g. 6.X.X) due to the change in the database schema?

Bruce

Reply:
> What is uPortal's process for updating the database schema to support a new release?

Whenever possible an automated migration path is provided. When an automated process is not possible a notice is added in the release notes describing the manual steps to migrate.

My understanding is import/export is frequently used for helping with migrations, @drewwills would be able to elaborate more on that.

---

> Do you feel this must be a major semantic version update (e.g. 6.X.X) due to the change in the database schema?

**short answer**
There is still ongoing discussion on if/how the DB schema included in the API.
Personally, I lean towards yes, which means this would be SemVer major (6.0).
I think this could be made SemVer minor by providing a default value if there is a `Null`.

**long answer**
uPortal is still in a monolithic repo, with monolithic versioning, which leads to some ambiguity around, [section 1 of the SemVer Spec](https://semver.org/spec/v2.0.0.html#semantic-versioning-specification-semver)

> Software using Semantic Versioning MUST declare a public API. This API could be declared in the code itself or exist strictly in documentation. However it is done, it should be precise and comprehensive.

uPortal hasn't finished clearly defining what is included as an API.

There is a colorable argument that the project is a monolith, that database is intended for internal usage only, and therefore is outside semantic versioning.
The is also a colorable argument that the subprojects in the monorepo are independent, each exposes its own API, whenever one of the subproject API's the Major version should be incremented.

I lean toward the former view. Meaning changes to the API that add a new required element to the API should be SemVer major.
Making the column optional could bring this back to SemVer minor.